chapters/synchronization.txt

```diff
@@ -53,67 +53,133 @@ _memory dependencies_ between two sets of operations defined by the
+operation must: complete before the second operation is initiated.
 More precisely:
 
-  * Let *A* and *B* be separate sets of operations.
+  * Let *Ops* be a set of all submitted operations.
   * Let *S* be a synchronization command.
-  * Let *A~S~* and *B~S~* be the synchronization scopes of *S*.
-  * Let *A'* be the intersection of sets *A* and *A~S~*.
-  * Let *B'* be the intersection of sets *B* and *B~S~*.
-  * Submitting *A*, *S* and *B* for execution, in that order, will result in
-    execution dependency *E* between *A'* and *B'*.
-  * Execution dependency *E* guarantees that *A'* happens-before *B'*.
+  * Let *Scope~src~* be the first synchronization scope of *S*.
+  * Let *Scope~dst~* be the second synchronization scope of *S*.
+  * Let *Ops~src~* be the subset of operations in *Ops* selected by scope
+    *Src~S~*.
+  * Let *Ops~dst~* be the subset of operations in *Ops* selected by scope
+    *Dst~S~*.
+  * Submitting *S* for execution will result in an execution dependency *ED*
+    between *Ops~src~* and *Ops~dst~*.
+  * Execution dependency *ED* guarantees that *Ops~src~* happens-before
+    *Ops~dst~*.
 
 [[synchronization-dependencies-chains]]
-An _execution dependency chain_ is a sequence of execution dependencies that
-form a happens-before relation between the first dependency's *A'* and the
-final dependency's *B'*.
-For each consecutive pair of execution dependencies, a chain exists if the
-intersection of *B~S~* in the first dependency and *A~S~* in the second
-dependency is not an empty set.
-The formation of a single execution dependency from an execution dependency
-chain can be described by substituting the following in the description of
-execution dependencies:
-
-  * Let *S* be a set of synchronization commands that generate an execution
-    dependency chain.
-  * Let *A~S~* be the first synchronization scope of the first command in
-    *S*.
-  * Let *B~S~* be the second synchronization scope of the last command in
-    *S*.
+A sequence of execution dependencies can: form a transitive property called
+an _execution dependency chain_.
+An _execution dependency chain_ is formed between any and all two execution
+dependencies whenever there is an overlap in the first dependency's
+*Scope~dst~* and subsequent dependency's *Scope~src~*.
+There is an overlap between two scopes if it is not possible to reduce the
+combination of the selection criteria of the *Scope~dst~* and the
+*Scope~src~* to nothing.
+
+[NOTE]
+.Note
+====
+For example, if a flink:vkCmdPipelineBarrier has pname:dstStageMask of
+ename:VK_PIPELINE_STAGE_FRAGMENT_SHADER_BIT, and subsequent
+flink:vkCmdPipelineBarrier has pname:srcStageMask of
+ename:VK_PIPELINE_STAGE_COLOR_ATTACHMENT_OUTPUT_BIT, the combined selection
+criterion of both is "commands occuring later than the first barrier and
+earlier than the second barrier reduced to the operations in
+ename:VK_PIPELINE_STAGE_FRAGMENT_SHADER_BIT,
+ename:VK_PIPELINE_STAGE_EARLY_FRAGMENT_TESTS_BIT,
```

Comment:
I think this is a really helpful example.

Just one question. I could be wrong about this, but I'm thinking VK_PIPELINE_STAGE_EARLY_FRAGMENT_TESTS_BIT might not be included in the combined selection criteria, because this stage happens before the fragment shader stage - does that seem correct to you?

Reply:
Copypasta error. I think they are listed this way in the enum, but in the graphics pipe there's just late test inbetween

Reply:
I'm not keen on this specific example because it leans on the pipeline stage order, which isn't defined yet. I'd rather we highlight a simpler example here to avoid new readers getting confused - e.g. something that includes exactly the same stages directly. We could then elaborate on this example later when pipeline stage order is introduced to get that concept highlighted better (which might be a good improvement regardless).

For us I think this example makes sense, because we've read the spec and are familiar - but for a new reader this would be confusing.

Reply:
I do think examples might make clear some of the abstract language of scopes and dependencies. So I encourage that.

Reply:
@Tobski I try to make as brutal example here as possible. I suppose it is sort of a desperate attempt to compensate the lack of sufficiently formal description of what "scope overlap" is...

In my mind the concept is pretty straightforward, so it is mindly annoying that it might probably require like half a page to properly explain in the end. And even so will probably require an example to have some certainty people grok it correct.

Reply:
@krOoze I don't disagree with your intent, but I do think a fresh reader would not be able parse this particular example until they've gone and read a lot more of the chapter. IMO we need something in the style of the Vulkan Guide or a proposal document - where you get a much higher level view of things without all the formality. There you could say a lot of this much more straightforwardly.

In the meantime, I'd rather we simplify this particular example.

Reply:
@Tobski I was kinda strategically aiming for something hard. This is the where reader needs to slow down. The example demonstrates the original Issue. I suppose I can sacrifice the implicit pipeline stage part of the example for now?

Reply:
Right, the pipeline stage order is really my main concern here - the rest is fine.